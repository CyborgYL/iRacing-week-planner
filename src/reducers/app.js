import moment, { duration } from 'moment';
import { CHANGE_MODAL, UPDATE_DAYS } from '../actions/app';
import { SIGNED_IN } from '../actions/auth';
import { seasonStart, weekSeasonStart } from '../config';

const currentDate = moment(new Date()).utc().startOf('day');
const currentDays = currentDate.diff(seasonStart, 'days');
const getWeek = (date) => Math.ceil(duration(moment(date).add({ second: 1 }).diff(weekSeasonStart)).asWeeks());

export default function app(
  state = {
    date: currentDate,
    daysSinceSeasonStart: currentDays,
    week: getWeek(currentDate),
    currentModal: null,
  },
  { type, days, modalName },
) {
  if (type === UPDATE_DAYS) {
    const date = moment(seasonStart).add(days, 'days');
    return {
      ...state,
      date,
      daysSinceSeasonStart: days,
      week: getWeek(date),
    };
  }

  if (type === CHANGE_MODAL) {
    return {
      ...state,
      currentModal: modalName,
    };
  }

  if (type === SIGNED_IN) {
    return {
      ...state,
      currentModal: null,
    };
  }

  return state;
}
