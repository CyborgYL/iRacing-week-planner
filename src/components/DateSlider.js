// @flow

import { Slider } from '@blueprintjs/core';
import * as React from 'react';
import { useTranslation } from 'react-i18next';
import { shallowEqual, useDispatch, useSelector } from 'react-redux';
import { updateDays } from '../actions/app';

import { seasonStart, seasonEnd } from '../config';
import styles from '../styles/main.module.scss';

const seasonLengthDays = seasonEnd.diff(seasonStart, 'days');

const dateDaysSelector = (state) => state.app.daysSinceSeasonStart;
const dateSelector = (state) => state.app.date;
const weekSelector = (state) => state.app.week;

export default function DateSlider() {
  const { t } = useTranslation();
  const dateDays = useSelector(dateDaysSelector);
  const date = useSelector(dateSelector, shallowEqual);
  const week = useSelector(weekSelector);
  const dispatch = useDispatch();

  return (
    <>
      <div className={styles.row}>
        <h3 className={styles['col-xs-8']}>
          {t('Races for date: {{date, YYYY MMM DD}}', { date: date.local().toDate() })}
        </h3>
        <h3 className={styles['col-xs-4']} style={{ textAlign: 'right' }}>
          {t('Week {{week}}', { week })}
        </h3>
      </div>
      <div style={{ marginBottom: 10 }}>
        <Slider
          min={0}
          max={seasonLengthDays}
          value={dateDays}
          stepSize={1}
          onChange={(days) => dispatch(updateDays(days))}
          labelRenderer={false}
        />
      </div>
    </>
  );
}
