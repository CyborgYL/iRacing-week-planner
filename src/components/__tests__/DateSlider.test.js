import { Slider } from '@blueprintjs/core';
import { describe, test } from '@jest/globals';
import moment from 'moment';
import React from 'react';
import renderer from 'react-test-renderer';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import { UPDATE_DAYS } from '../../actions/app';

import DateSlider from '../DateSlider';

const mockStore = configureMockStore([thunk]);

describe('components/DateSlider', () => {
  const defaultStore = {
    app: {
      date: moment('2020-09-23T01:00:00.000Z'),
      daysSinceSeasonStart: 15,
      week: 3,
    },
  };

  test('renders correctly', () => {
    const store = mockStore(defaultStore);
    const component = renderer.create(<Provider store={store}><DateSlider /></Provider>);

    expect(component.toJSON()).toMatchSnapshot();

    component.root.findByType(Slider).props.onChange(3);

    expect(store.getActions()[0].type).toEqual(UPDATE_DAYS);
    expect(store.getActions()[0].days).toBe(3);
  });
});
