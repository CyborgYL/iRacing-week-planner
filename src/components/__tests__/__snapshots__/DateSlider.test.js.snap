// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/DateSlider renders correctly 1`] = `
Array [
  <div
    className="row"
  >
    <h3
      className="col-xs-8"
    >
      Races for date: 2020 Sep 23
    </h3>
    <h3
      className="col-xs-4"
      style={
        Object {
          "textAlign": "right",
        }
      }
    >
      Week 3
    </h3>
  </div>,
  <div
    style={
      Object {
        "marginBottom": 10,
      }
    }
  >
    <div
      className="bp3-slider bp3-slider-unlabeled"
      onMouseDown={[Function]}
      onTouchStart={[Function]}
    >
      <div
        className="bp3-slider-track"
      >
        <div
          className="bp3-slider-progress"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
        <div
          className="bp3-slider-progress bp3-intent-primary"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
        <div
          className="bp3-slider-progress"
          style={
            Object {
              "left": "0.00%",
              "right": "100.00%",
              "top": 0,
            }
          }
        />
      </div>
      <div
        className="bp3-slider-axis"
      />
      <span
        className="bp3-slider-handle"
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onTouchStart={[Function]}
        style={
          Object {
            "left": "calc(0.00% - 0px)",
          }
        }
        tabIndex={0}
      />
    </div>
  </div>,
]
`;
